pkg/monitor/worker.go

```diff
@@ -0,0 +1,168 @@
+package monitor
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+import (
+	"context"
+	"strconv"
+	"time"
+
+	"github.com/sirupsen/logrus"
+	"k8s.io/client-go/kubernetes"
+
+	"github.com/Azure/ARO-RP/pkg/api"
+	"github.com/Azure/ARO-RP/pkg/util/recover"
+	"github.com/Azure/ARO-RP/pkg/util/restconfig"
+)
+
+// listBuckets reads our bucket allocation from the master
+func (mon *monitor) listBuckets(ctx context.Context) error {
+	buckets, err := mon.db.Monitors.ListBuckets(ctx)
+	mon.baseLog.Printf("servicing %d buckets", len(buckets))
+
+	mon.mu.Lock()
+	defer mon.mu.Unlock()
+
+	mon.buckets = map[int]struct{}{}
+
+	if err != nil {
+		return err
+	}
+
+	for _, i := range buckets {
+		mon.buckets[i] = struct{}{}
+	}
+
+	return nil
+}
+
+// filler tracks the OpenShiftClusters change feed and keeps mon.docs
+// up-to-date.  We don't monitor clusters in Creating state, hence we don't add
+// them to mon.docs.  We also don't monitor clusters in Deleting state; when
+// this state is reached we delete from mon.docs
+func (mon *monitor) filler(ctx context.Context, log *logrus.Entry, stop <-chan struct{}) {
+	defer recover.Panic(log)
+
+	i := mon.db.OpenShiftClusters.ChangeFeed()
```

Comment:
Does the changeFeed give all "previous" documents when it starts or just ones created/modified during the time we are listening?

Reply:
It is like list and watch in one go.  It brings you up-to-date from zero then gives you subsequent modifications.  I found  https://medium.com/@gary.strange/cosmosdb-change-feed-processing-c002448a6ca really useful on this.